File: src/java/foundgine-core/src/main/java/com/foundgine/core/semantic/SemanticValue.java
package com.foundgine.core.semantic;

import java.math.BigDecimal;
import java.time.*;
import java.util.*;

/** Provider-independent canonical semantic value. */
public record SemanticValue(SemanticValueKind kind, Object value) {
    public static final SemanticValue NULL = new SemanticValue(SemanticValueKind.NULL, null);
    public static SemanticValue from(Object value) {
        if (value == null) return NULL;
        if (value instanceof String || value instanceof Character) return new SemanticValue(SemanticValueKind.STRING, value.toString());
        if (value instanceof Boolean) return new SemanticValue(SemanticValueKind.BOOLEAN, value);
        if (value instanceof Byte || value instanceof Short || value instanceof Integer || value instanceof Long)
            return new SemanticValue(SemanticValueKind.INT64, ((Number)value).longValue());
        if (value instanceof Float || value instanceof Double || value instanceof BigDecimal)
            return new SemanticValue(SemanticValueKind.DECIMAL, new BigDecimal(value.toString()));
        if (value instanceof Instant || value instanceof LocalDateTime || value instanceof OffsetDateTime || value instanceof ZonedDateTime || value instanceof LocalDate || value instanceof java.util.Date)
            return new SemanticValue(SemanticValueKind.DATETIME, value);
        if (value instanceof UUID) return new SemanticValue(SemanticValueKind.GUID, value);
        if (value instanceof Enum<?>) return new SemanticValue(SemanticValueKind.ENUM, value.toString());
        if (value instanceof Iterable<?> it) { List<SemanticValue> list = new ArrayList<>(); it.forEach(v -> list.add(from(v))); return new SemanticValue(SemanticValueKind.LIST, List.copyOf(list)); }
        return new SemanticValue(SemanticValueKind.OBJECT, value.toString());
    }
    @Override public String toString() { return kind == SemanticValueKind.NULL ? "null" : String.valueOf(value); }
}

File: src/csharp/Foundgine.Core/Semantic/SemanticValue.cs
namespace Foundgine.Core.Semantic;

/// <summary>
/// Provider-independent, canonical semantic value. Adapters may still accept
/// CLR objects at their boundary, but semantic validation and canonicalization
/// can operate on this representation without depending on a provider type.
/// </summary>
public readonly record struct SemanticValue(SemanticValueKind Kind, object? Value)
{
    public static SemanticValue Null => new(SemanticValueKind.Null, null);

    public static SemanticValue From(object? value)
    {
        if (value is null) return Null;
        return value switch
        {
            string or char => new(SemanticValueKind.String, value.ToString()),
            bool v => new(SemanticValueKind.Boolean, v),
            byte or sbyte or short or ushort or int or uint or long or ulong => new(SemanticValueKind.Int64,
                Convert.ToInt64(value, System.Globalization.CultureInfo.InvariantCulture)),
            float or double or decimal => new(SemanticValueKind.Decimal,
                Convert.ToDecimal(value, System.Globalization.CultureInfo.InvariantCulture)),
            DateTime v => new(SemanticValueKind.DateTime, v.ToUniversalTime()),
            DateTimeOffset v => new(SemanticValueKind.DateTime, v.ToUniversalTime()),
            Guid v => new(SemanticValueKind.Guid, v),
            Enum v => new(SemanticValueKind.Enum, v.ToString()),
            System.Collections.IEnumerable values when value is not string => new(SemanticValueKind.List,
                values.Cast<object?>().Select(From).ToArray()),
            _ => new(SemanticValueKind.Object, value.ToString())
        };
    }

    public override string ToString() => Kind switch
    {
        SemanticValueKind.Null => "null",
        SemanticValueKind.List =>
            $"[{string.Join(',', ((IReadOnlyList<SemanticValue>)Value!).Select(x => x.ToString()))}]",
        _ => Value?.ToString() ?? "null"
    };
}

public enum SemanticValueKind : byte
{
    Null,
    String,
    Boolean,
    Int64,
    Decimal,
    DateTime,
    Guid,
    Enum,
    List,
    Object
}